### PR TITLE
Change Application<>Company association to has_and_belongs_to_many

### PR DESCRIPTION
**Pull request by patmbolger**
PT Story: https://www.pivotaltracker.com/story/show/154143904

This is an initial step in converting our models to allow an ShfApplication to be associated with many companies.   There are other stories that will continue this effort in the epic: https://www.pivotaltracker.com/story/show/154143904

Changes proposed in this pull request:
1. Change app-to-company association from "belongs_to" to "has_and_belongs_to_many"
2. Change company-to-app association from "has_many" to "has_and_belongs_to_many"
3. Introduce short-term changes in code to accommodate this change

Ready for review:
@AgileVentures/shf-project-team

### CODE REVIEW

File: lib/tasks/one_time/2018_Q1/app_has_many_companies.rake [hunk 1]
```diff
@@ -0,0 +1,27 @@
+namespace :shf do
+
```
Comment (weedySeaDragon) [inline lib/tasks/one_time/2018_Q1/app_has_many_companies.rake]: Can we insert a (sub-) namespace `onetime` to help separate these out?  Also makes it much less likely that we will accidentally run it due to file/line/text completion.

File: lib/tasks/one_time/2018_Q1/app_has_many_companies.rake [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  desc "Convert has_one data to has_many association"
+  task app_has_many_companies: :environment do
+
+    log_file = 'log/app_has_many_companies'
+
+    ActivityLogger.open(log_file, 'App>>Company', 'convert to has_many') do |log|
+
+      apps_converted = 0
+      log.record('info', "Checking #{ShfApplication.count} applications.")
```
Comment (weedySeaDragon) [inline lib/tasks/one_time/2018_Q1/app_has_many_companies.rake]: For anyone looking at the log at a later date, who hasn't seen (or has forgotten) the code, it would be helpful to include text that explains why the number of applications converted might be < the number "Checking".  Suggestion:
`"Checking #{ShfApplication.count} applications. Only those with an associated company_id will be converted."`

File: app/mailers/shf_application_mailer.rb [hunk 3]
```diff
@@ -9,7 +9,7 @@ def acknowledge_received(shf_application)
 
 
   def app_approved(shf_application)
-    @branding_fee_paid = shf_application.company&.branding_license?
+    @branding_fee_paid = shf_application.companies.last&.branding_license?
```
Comment (weedySeaDragon) [inline app/mailers/shf_application_mailer.rb:12]: This isn't correct. Can't just check the last company.  What if some companies have the branding fee paid but some don't?
Probably should change the name of this instance var to "@all_branding_fees_paid" to communicate the intent clearly.  (Don't forget to change it in the view(s), etc.)

(So obviously you'll need to change the _specifications_ in `spec/mailers/shf_application_mailer_spec.rb` and test for cases where some are paid, some aren't, etc.)

This is someplace where we can't leave this in a half-way state.
Because this is a time when many people are making payments and we don't want to make things any more confusing than necessary, I think we should make this change now in this PR.

File: spec/models/shf_application_spec.rb [hunk 4]
```diff
@@ -163,7 +163,7 @@
     let(:application2) do
       create(:shf_application, user: user2,
              uploaded_files: [uploaded_file], state: :new,
-             company_id: application.company.id,
+             companies: [application.companies.last],
              company_number: application.company_number)
```
Comment (weedySeaDragon) [inline spec/models/shf_application_spec.rb:167]: What are we going to do about `company_number` ???
Seems that this should no longer be an attribute of a SHF Application, right?
So should we also remove it in this PR?  (including removing it from all views)

File: spec/models/shf_application_spec.rb [hunk 5]
```diff
@@ -172,14 +172,9 @@
       expect(uploaded_file.destroyed?).to be_truthy
     end
 
-    it "destroys associated company if it has no remaining applications" do
```
Comment (weedySeaDragon) [inline spec/models/shf_application_spec.rb:175]: Why did you remove this?  Is it because we need to have the explicit discussion with the client about this?  (I'm assuming so, but just wanted to be sure.)

File: spec/models/shf_application_spec.rb [hunk 6]
```diff
@@ -358,7 +353,9 @@
 
     it 'uses the company main address' do
 
-      expect(accepted_app.se_mailing_csv_str).to eq AddressExporter.se_mailing_csv_str(accepted_app.company.main_address)
+      expect(accepted_app.se_mailing_csv_str)
+        .to eq AddressExporter
+        .se_mailing_csv_str(accepted_app.companies.last.main_address)
```
Comment (weedySeaDragon) [inline spec/models/shf_application_spec.rb:358]: If this is ok with @thesuss  for now, then I'm ok with this (and the related code that satisfies this), _provided that_ you please change the description to something that makes this explicit so we don't forget to change this in the future.  
Suggestion:
`it 'uses the main address of the last company in the list of companies' do`

File: spec/models/shf_application_spec.rb [hunk 7]
```diff
@@ -338,7 +333,7 @@
           application.accept!
         end
         it 'assigns company email to application contact_email' do
-          expect(application.company.email).to eq application.contact_email
+          expect(application.companies.last.email).to eq application.contact_email
```
Comment (weedySeaDragon) [inline spec/models/shf_application_spec.rb]: Update the specification to accurately reflect the change, please.

Suggestion:
`it 'uses the email from the last company in the list for the application contact_email default' do`

File: app/controllers/admin_controller.rb [hunk 8]
```diff
@@ -44,7 +44,7 @@ def export_str(shf_apps)
       out_str << ','
 
       # a company name may have commas, so surround with quotes so spreadsheets recognize it as one string and not multiple comma-separated value
-      out_str << (m_app.company.nil? ?  '' : "\"#{m_app.company.name}\"")
+      out_str << (m_app.companies.empty? ?  '' : "\"#{m_app.companies.last.name}\"")
```
Comment (weedySeaDragon) [inline app/controllers/admin_controller.rb:47]: OK with me if this is OK with @thesuss .

File: spec/controllers/admin_controller_spec.rb [hunk 9]
```diff
@@ -94,7 +94,7 @@
             result_str << member1_info + ','
             result_str << '"",'  # no business categories
 
-            result_str << (m.company.nil? ? '' : '"' + m.company.name + '"')
+            result_str << (m.companies.any? ? '"' + m.companies.last.name + '"' : '')
```
Comment (weedySeaDragon) [inline spec/controllers/admin_controller_spec.rb:97]: OK with me if it's OK with @thesuss .

It would be _best_ to make this specification explicit by having it be a separate spec/example.  I'll make a note in the PT feature that is outstanding that is about changing this code.  We can refactor this spec in that PR.

**Review by weedySeaDragon — CHANGES_REQUESTED** [review]
see comments &  what Houndbot says (to be consistent with Ruby style and boolean methods)

File: app/controllers/admin_controller.rb (hunk 8 above)
Comment (patmbolger) [reply app/controllers/admin_controller.rb:47]: Or, we could list a comma-separated list of company names for the application(?)

I did not go that way because I am not sure if that is important for this exported list of apps.

File: app/mailers/shf_application_mailer.rb (hunk 3 above)
Comment (patmbolger) [reply app/mailers/shf_application_mailer.rb:12]: I did not change this because:

1. We currently only support the ability of a user to associate one company with  an application (in the UI).  So, we _shouldn't_ see a situation involving more than one company until we relax that constraint.  Whether supporting this case would involve a re-review of the application is not decided.

2. _That_ situation might arise when we figure out how to handle application renewals and/or adding a category to an existing application (at which point it is very possible to have more than one company associated with the app).  But, we don't support that now either.

My thinking on this is that this email would probably go out _once_ to the user - when their application is first accepted.  

Case #1 above would probably require another email type altogether - that is, we're not reviewing the application but simply associating that app with another company (and, if the member can do this themselves, no email is required).

Case #2 would almost certainly require a different email (or emails - one for bi-annual review completion and one for adding one or more categories).

So, I had the same thoughts as you when I adjusted this code.  Given my thoughts above, however, I decided to leave it alone for now - as we think through these refactoring side-effects which are very fundamental to how we model the SHF world.

I also think that the best way to deal with this is to provide this email action with the specific company involved with the initial application, so it doesn't have to perform this seemingly crazy logic.

I'll create a story for that if you agree?  Thoughts?

File: lib/tasks/one_time/2018_Q1/app_has_many_companies.rake (hunk 1 above)
Comment (patmbolger) [reply lib/tasks/one_time/2018_Q1/app_has_many_companies.rake]: Good idea.  Now it will be:
```
rake shf:one_time:app_has_many_companies
```

File: lib/tasks/one_time/2018_Q1/app_has_many_companies.rake (hunk 2 above)
Comment (patmbolger) [reply lib/tasks/one_time/2018_Q1/app_has_many_companies.rake]: Right .... I added more info to the log.  It now looks like:
```
[App>>Company] [convert to has_many] [info] Started at 2018-01-19 14:15:56 UTC
[App>>Company] [convert to has_many] [info] This one-time conversion is associated with changing:
[App>>Company] [convert to has_many] [info] >> the association between an SHF application
[App>>Company] [convert to has_many] [info] >> and companies from has_one to has_and_belongs_to_many.
[App>>Company] [convert to has_many] [info] Checking 0 applications:
[App>>Company] [convert to has_many] [info] >> (only applications with a company will be converted)
[App>>Company] [convert to has_many] [info] Converted 0 applications.
[App>>Company] [convert to has_many] [info] Finished at 2018-01-19 14:15:56 UTC.
[App>>Company] [convert to has_many] [info] Duration: 0.11 seconds.
```
(I ran it against an empty DB this time, but the pre-convert records are shown).

File: spec/models/shf_application_spec.rb (hunk 4 above)
Comment (patmbolger) [reply spec/models/shf_application_spec.rb:167]: @RobertCram has a story underway for that.  I added a comment to that story in the course of working on this PR.  (we may want to keep an attribute in the application - useful while it is under review and possible later for reporting purposes - but renamed to reflect its actual usage).

File: spec/models/shf_application_spec.rb (hunk 5 above)
Comment (patmbolger) [reply spec/models/shf_application_spec.rb:175]: It's not really due to the reason you mention (although we'd probably want to do this here, anyway, for that reason).

My _intent_ for this PR was to not change any unrelated logic.

However - _if_ we retained the deletion of the company when the app is deleted - than we will have to adopt another approach altogether (see story I created for this - https://www.pivotaltracker.com/story/show/154296716).  So, if we did this, it is a larger story than would fit into to this story.

(the prior way of destroying the company - when the app is detroyed - involved some rather tortuous logic between the two models - which is no longer viable with the change to `has_and_belongs_to_many` association.)

File: spec/models/shf_application_spec.rb (hunk 7 above)
Comment (patmbolger) [reply spec/models/shf_application_spec.rb]: For now I changed it to:
```
"assigns app's latest-added-company email to application contact_email"
```
but I did not change this originally because of how I think the need to refactor the "app accepted" email - see discussion above.

File: spec/models/shf_application_spec.rb (hunk 6 above)
Comment (patmbolger) [reply spec/models/shf_application_spec.rb:358]: now:
```
"uses the app's latest-added-company main address"
```

File: spec/models/shf_application_spec.rb (hunk 4 above)
Comment (patmbolger) [reply spec/models/shf_application_spec.rb:167]: Robert's story: https://www.pivotaltracker.com/story/show/151611254

**Review by weedySeaDragon — APPROVED** [review]

**Review by thesuss — APPROVED** [review]
